### PR TITLE
Share the love

### PR DESCRIPTION
**Pull request by dmdmdm**
Previously the Love button only worked for Lastfm play.
I modified it to work for local files too.
When the user clicks, they get a small dialog with the name of the playing track and a rating control.
Clicking on the rating (eg 4 stars of out 5) will:
- Save it in the mp3 file 
- Update the display in the current playlist (if has rating column)
- Close the dialog

### CODE REVIEW

File: src/ui/lovedialog.cpp [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  // Restore the tab that was current last time.
+  QSettings s;
+  s.beginGroup(kSettingsGroup);
+
+  QDialog::showEvent(e);
+}
+
+void LoveDialog::SongRated(float rating) {
+  qLog(Debug) << "LoveDialog::SongRated: rating=" << rating;
+  Song* song = &song_;
```
Comment (hatstand) [inline src/ui/lovedialog.cpp:71]: I don't think you need this variable, just use `song_` directly.

File: src/ui/lovedialog.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+   along with Clementine.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef LOVE_DIALOG_H
+#define LOVE_DIALOG_H
+
+#include <QDialog>
+
+#include "core/song.h"
+//#include "widgets/lineedit.h"
```
Comment (hatstand) [inline src/ui/lovedialog.h:24]: Remove

File: src/ui/mainwindow.cpp [hunk 3]
```diff
@@ -1582,13 +1586,28 @@ void MainWindow::ScrobbledRadioStream() {
   ui_->action_love->setEnabled(true);
   if (tray_icon_) tray_icon_->LastFMButtonLoveStateChanged(true);
 }
+#endif
 
 void MainWindow::Love() {
-  app_->scrobbler()->Love();
-  ui_->action_love->setEnabled(false);
-  if (tray_icon_) tray_icon_->LastFMButtonLoveStateChanged(false);
+  Playlist* activePlaylist = app_->playlist_manager()->active();
```
Comment (hatstand) [inline src/ui/mainwindow.cpp:1592]: `active_playlist`

File: src/ui/mainwindow.cpp [hunk 4]
```diff
@@ -1582,13 +1586,28 @@ void MainWindow::ScrobbledRadioStream() {
...
+    qLog(Warning) << "Love: nothing playing so can't love it";
+    return;
+  }
+
+  if (item->IsLocalLibraryItem()) {
+    Song song = item->Metadata();
+    love_dialog_->SetSong(song);
+    love_dialog_->show();
+  }
+  else {
```
Comment (hatstand) [inline src/ui/mainwindow.cpp:1605]: Same line as above

File: src/ui/mainwindow.cpp [hunk 5]
```diff
@@ -1582,13 +1586,28 @@ void MainWindow::ScrobbledRadioStream() {
...
+  Playlist* activePlaylist = app_->playlist_manager()->active();
+  PlaylistItemPtr item = activePlaylist->current_item();
+  if (!item) {
+    // Don't make a big deal about it
+    qLog(Warning) << "Love: nothing playing so can't love it";
+    return;
+  }
+
+  if (item->IsLocalLibraryItem()) {
+    Song song = item->Metadata();
```
Comment (hatstand) [inline src/ui/mainwindow.cpp:1601]: `const Song& song`

File: src/ui/lovedialog.cpp [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+  QSettings s;
+  s.beginGroup(kSettingsGroup);
+
+  QDialog::showEvent(e);
+}
+
+void LoveDialog::SongRated(float rating) {
+  qLog(Debug) << "LoveDialog::SongRated: rating=" << rating;
+  qLog(Debug) << "LoveDialog::SongRated: isValid=" << song_.is_valid() << " id=" << song_.id();
+  if (!song_.is_valid() || song_.id() == -1) return;
```
Comment (jbroadus) [inline src/ui/lovedialog.cpp:72]: This is done elsewhere, but I think it's actually broken. From what I saw, a non-library song from a playlist will be given the id from the playlist table. So, the rating would be set for the wrong item.

That said, this is no worse than what the tag editing dialog is currently doing. I was testing a changeset that set a library flag in the song metadata. I'll dig that out and see if it's ready for review.

Comment (dmdmdm) [reply src/ui/lovedialog.cpp:72]: In mainwindow.cpp I do `if (item->IsLocalLibraryItem()) { ...` before sending the user to LoveDialog.  So it should be a library song in LoveDialog::SongRated() or that's what I think ;).  I look forward to hearing how your testing went.  I want this to work!

Comment (jbroadus) [reply src/ui/lovedialog.cpp:72]: Yep.  IsLocalLibraryItem() should insure that you're using the correct backend.

**Comment by jonaski** [comment]
I'm surprised this wasn't stopped. It's now impossible to love songs that are part of the library, for what purpose? The ratings are easily accessible from the playlist, if there should be an additional dialog for rating it should be a separate button with it's own setting, at least not replace loving songs when both last.fm and the love button is enabled in settings.

**Comment by dmdmdm** [comment]
I'm at a place with really limited internet. The reason for the change:
rating control isn't in the playlist by default, as a beginner user I
clicked on the heart for a local track and got an error about last.fm -
very confusing and unfriendly. My intention was to make local rating easy
without breaking last.fm love. But sounds like it broke a bit. I can
improve it next week. Thanks for the feedback

**Comment by Echocs7** [comment]
Can I understand why `MainWindow::Love()` is outside `#ifdef HAVE_LIBLASTFM` while it is still inside `#ifdef HAVE_LIBLASTFM` in mainwindow.h ?

**Pull request merged**